dial_queue.go

```diff
@@ -0,0 +1,269 @@
+// end up adding fuel to the fire. Since we have no deterministic way to detect this for now, we hard-limit concurrency
+// to DialQueueMaxParallelism.
+func newDialQueue(ctx context.Context, target string, in *queue.ChanQueue, dialFn func(context.Context, peer.ID) error, nConsumers int) *dialQueue {
+	sq := &dialQueue{
+		ctx:           ctx,
+		dialFn:        dialFn,
+		nWorkers:      DialQueueMinParallelism,
+		scalingFactor: 1.5,
+
+		in:  in,
+		out: queue.NewChanQueue(ctx, queue.NewXORDistancePQ(target)),
+
+		growCh:    make(chan struct{}, nConsumers),
+		shrinkCh:  make(chan struct{}, 1),
+		waitingCh: make(chan waitingCh, nConsumers),
+		dieCh:     make(chan struct{}, DialQueueMaxParallelism),
+	}
+	for i := 0; i < DialQueueMinParallelism; i++ {
+		go sq.worker()
+	}
+	go sq.control()
+	return sq
+}
+
+func (dq *dialQueue) control() {
+	var (
+		p              peer.ID
+		dialled        = dq.out.DeqChan
+		resp           waitingCh
+		waiting        <-chan waitingCh
+		lastScalingEvt = time.Now()
+	)
+	for {
+		// First process any backlog of dial jobs and waiters -- making progress is the priority.
+		// This block is copied below; couldn't find a more concise way of doing this.
+		select {
+		case <-dq.ctx.Done():
+			return
+		case p = <-dialled:
+			dialled, waiting = nil, dq.waitingCh
+			continue // onto the top.
+		case resp = <-waiting:
+			// got a channel that's waiting for a peer.
+			log.Debugf("delivering dialled peer to DHT; took %dms.", time.Now().Sub(resp.ts)/time.Millisecond)
+			resp.ch <- p
+			close(resp.ch)
+			dialled, waiting = dq.out.DeqChan, nil // stop consuming waiting jobs until we've cleared a peer.
+			continue                               // onto the top.
+		default:
+			// there's nothing to process, so proceed onto the main select block.
+		}
+
+		select {
+		case <-dq.ctx.Done():
+			return
+		case p = <-dialled:
+			dialled, waiting = nil, dq.waitingCh
+		case resp = <-waiting:
+			// got a channel that's waiting for a peer.
+			log.Debugf("delivering dialled peer to DHT; took %dms.", time.Now().Sub(resp.ts)/time.Millisecond)
+			resp.ch <- p
+			close(resp.ch)
+			dialled, waiting = dq.out.DeqChan, nil // stop consuming waiting jobs until we've cleared a peer.
+		case <-dq.growCh:
+			if time.Now().Sub(lastScalingEvt) < DialQueueScalingMutePeriod {
+				continue
+			}
+			dq.grow()
+			lastScalingEvt = time.Now()
+		case <-dq.shrinkCh:
+			if time.Now().Sub(lastScalingEvt) < DialQueueScalingMutePeriod {
+				continue
+			}
+			dq.shrink()
+			lastScalingEvt = time.Now()
+		}
+	}
+}
+
+func (dq *dialQueue) Consume() (<-chan peer.ID, error) {
+	ch := make(chan peer.ID, 1)
+
+	// short circuit and return a dialled peer if it's immediately available.
+	select {
+	case <-dq.ctx.Done():
+		return nil, ErrContextClosed
+	case p := <-dq.out.DeqChan:
+		ch <- p
+		close(ch)
+		return ch, nil
+	default:
+	}
+
+	// we have no finished dials to return, trigger a scale up.
+	select {
+	case dq.growCh <- struct{}{}:
```

Comment:
May not be the bet API but it should be considered: Instead of explicitly growing like this, we could:

1. Expose a channel to the user.
2. When we have no new connections, try writing `nil` to the channel.
3. If a consumer reads `nil`, (a) it should throw it away and try again and (b) we should treat that as a signal to grow.

We could also just have the consumer read from *two* channels.

---

Just something to consider. The current API is definitely *cleaner*.

Reply:
I did consider consumer-side signalling, but I didn't want to introduce that level of coupling. This fix is really the first iteration of this queue (which could be exportable to other areas of libp2p); we might need to get more sophisticated over time, and making the consumer conform to a specific choreography can be very limiting and end up becoming technical debt in the future.

Reply:
SGTM.